runtime/interface.go

```diff
@@ -41,55 +43,130 @@ type Block struct {
+	// Ramtin: (we might not need this to be passed and just be enforced in the Results)
+	ComputationLimit() uint64
+}
+
+// CacheProvider provides caching functionality to the cadence runtime
+//
+// cache should be tx failure aware (rollback changes)
+// cache should not break verification (register touches when returns)
+type CacheProvider interface {
 	// GetCachedProgram attempts to get a parsed program from a cache.
 	GetCachedProgram(Location) (*ast.Program, error)
 	// CacheProgram adds a parsed program to a cache.
 	CacheProgram(Location, *ast.Program) error
-	// GetValue gets a value for the given key in the storage, owned by the given account.
-	GetValue(owner, key []byte) (value []byte, err error)
-	// SetValue sets a value for the given key in the storage, owned by the given account.
-	SetValue(owner, key, value []byte) (err error)
-	// CreateAccount creates a new account.
-	CreateAccount(payer Address) (address Address, err error)
-	// AddAccountKey appends a key to an account.
-	AddAccountKey(address Address, publicKey []byte) error
-	// RemoveAccountKey removes a key from an account by index.
-	RemoveAccountKey(address Address, index int) (publicKey []byte, err error)
-	// UpdateAccountContractCode updates the code associated with an account contract.
-	UpdateAccountContractCode(address Address, name string, code []byte) (err error)
-	// GetAccountContractCode returns the code associated with an account contract.
-	GetAccountContractCode(address Address, name string) (code []byte, err error)
-	// RemoveAccountContractCode removes the code associated with an account contract.
-	RemoveAccountContractCode(address Address, name string) (err error)
-	// GetSigningAccounts returns the signing accounts.
-	GetSigningAccounts() ([]Address, error)
-	// Log logs a string.
-	Log(string) error
-	// EmitEvent is called when an event is emitted by the runtime.
-	EmitEvent(cadence.Event) error
-	// ValueExists returns true if the given key exists in the storage, owned by the given account.
-	ValueExists(owner, key []byte) (exists bool, err error)
-	// GenerateUUID is called to generate a UUID.
-	GenerateUUID() (uint64, error)
-	// GetComputationLimit returns the computation limit. A value <= 0 means there is no limit
-	GetComputationLimit() uint64
-	// SetComputationUsed reports the amount of computation used.
-	SetComputationUsed(used uint64) error
-	// DecodeArgument decodes a transaction argument against the given type.
-	DecodeArgument(argument []byte, argumentType cadence.Type) (cadence.Value, error)
-	// GetCurrentBlockHeight returns the current block height.
-	GetCurrentBlockHeight() (uint64, error)
-	// GetBlockAtHeight returns the block at the given height.
-	GetBlockAtHeight(height uint64) (block Block, exists bool, err error)
-	// UnsafeRandom returns a random uint64, where the process of random number derivation is not cryptographically
-	// secure.
-	UnsafeRandom() (uint64, error)
+}
+
+type CryptoProvider interface {
```

Comment:
I put this one back in for now, I thought we might actually need hash functionality for some cadence operations and it was a great deal of dependency that we can figure out later.

Reply:
Yes, I think let's keep this in for now and then remove it in a subsequent PR by replacing them with injected functions.